myjsonrpc.pm

```diff
@@ -53,9 +54,11 @@ sub send_json {
 
     my $wb = syswrite($to_fd, "$json");
     if (!$wb || !$wb == length($json)) {
-        confess "syswrite failed: $!" if DEBUG_JSON;
-        # remote end most likely terminated on request, silently ignore
-        return undef;
+        if (!DEBUG_JSON && $! =~ qr/Broken pipe/) {
+            bmwqemu::diag("myjsonrpc: remote end terminated connection, stopping");
+            _exit(0);
```

Comment:
careful here. if the backend is dead, we still don't want to kill isotovideo

Reply:
yes but I don't think that happens. I did not paste the full log in the commit message. It actually ends with:

```
[2019-11-22T09:50:24.694 CET] [debug] done with autotest process
[2019-11-22T09:50:24.694 CET] [debug] isotovideo failed
[2019-11-22T09:50:24.695 CET] [debug] stopping backend process 25141
[2019-11-22T09:50:24.695 CET] [debug] done with backend process
25133: EXIT 1
```

so, are we good?

Reply:
Not sure, I needed to do some more own tests (e.g. killing the command server while the test is executed. Isotovideo should then still perform the usual cleanup and not just exit). Because this really looks like isotovideo would now just stop when a broken pipe is encountered which is definitely wrong. And even if that's not the case a comment stating why this code is not as dangerous as one would assume might be helpful.

Reply:
I'm not saying it happens in your case. but myjsonrpc is used everywhere

Reply:
even if it's not dangerous now - it might be used in different places where this exit then becomes dangerous. So returning undef and exiting from the non-dangerous caller is still preferred.

Reply:
I see. I will try to do that. But please keep in mind that before 449baece it called "confess" in all cases so I would not see how _exit could be worse

Reply:
That is actually a good point. Maybe it works because everything is wrapped within the `if (!$wb || !$wb == length($json))` block. Not sure how to read that condition, though. The `!$wb == length($json)` part seems pointless because `$wb` is negated before the comparison. And that has changed with commit 449baece.

Reply:
I think I have found a good testing procedure now:

* start `isotovideo -d`, e.g. with os-autoinst-distri-opensuse where the tests would go into an `assert_screen` waiting for a screen for a bit longer
* as soon as all processes are running and the test is waiting for needle matches, e.g. "no match: 79.0s, best candidate: grub2-2-20131030 (0.00)" kill one of the processes and look for proper shutdown in each case, e.g.
  * `pkill -f "qemu-system-x86_64"`
  * `pkill -f "commands"`
  * `pkill -f "autotest"`
  * `pkill -f "backend"`

It seems with my change the shutdown in case of `pkill -f "commands"` leaves subprocesses behind that only end themselves after the main process exited, i.e. there is a bash prompt already but then you see output.

Reply:
And that might be because the main process (= `isotovideo`) exits without doing the cleanup code which would take care of the other processes. Have you tested this without your change as well?

Besides, it would be good if the messed up condition could be fixed as well.

Reply:
Yes, sure. I crosschecked "old" vs. "new". What "messed up conition"?

Reply:
@okurz `if (!$wb || !$wb == length($json))`

Reply:
Yes, that's the condition I've mean (and pointed out in a previous comment).